File: python_implementation/src/application/use_cases/tasks/list_tasks_use_case.py
from typing import Optional
from src.domain.abstractions.repositories.abstract_task_repository import AbstractTaskRepository
from src.domain.entities.task_entity import TaskEntity
from src.domain.enums.task_enums import TaskStatus
from src.config.constants import limit_cap, limit_default, offset_default


class ListTasksUseCase:
    """Use case for listing tasks with optional filtering"""
    
    def __init__(self, task_repository: AbstractTaskRepository):
        self.task_repository = task_repository
    
    def execute(
        self, 
        status: Optional[TaskStatus] = None,
        limit: int = limit_default,
        offset: int = offset_default
    ) -> tuple[list[TaskEntity], int]:
        # Validate and cap limit
        if limit > limit_cap:
            limit = limit_cap
        if limit < 1:
            limit = limit_default
        
        # Validate offset
        if offset < 0:
            offset = offset_default
        
        # Get tasks (repository will handle filtering if needed)
        tasks = self.task_repository.get_all(skip=offset, limit=limit, status=status)
        
        # Get total count for the filter
        total = self.task_repository.count(status=status)
        
        return tasks, total

File: java_implementation/src/main/java/com/example/taskapi/application/usecases/ListTasksUseCase.java
package com.example.taskapi.application.usecases;

import com.example.taskapi.config.Constants;
import com.example.taskapi.domain.enums.TaskStatus;
import com.example.taskapi.domain.model.Task;
import com.example.taskapi.domain.repository.TaskRepositoryPort;
import lombok.AllArgsConstructor;
import lombok.Data;
import lombok.RequiredArgsConstructor;
import org.springframework.stereotype.Service;

import java.util.List;

@Service
@RequiredArgsConstructor
public class ListTasksUseCase {
    private final TaskRepositoryPort taskRepository;

    @Data
    @AllArgsConstructor
    public static class ListTasksResult {
        private List<Task> tasks;
        private long total;
    }

    public ListTasksResult execute(TaskStatus status, Integer limit, Integer offset) {
        if (limit == null) {
            limit = Constants.LIMIT_DEFAULT;
        }
        if (limit > Constants.LIMIT_CAP) {
            limit = Constants.LIMIT_CAP;
        }
        if (limit < 1) {
            limit = Constants.LIMIT_DEFAULT;
        }

        if (offset == null || offset < 0) {
            offset = Constants.OFFSET_DEFAULT;
        }

        List<Task> tasks = taskRepository.getAll(offset, limit, status);
        long total = taskRepository.count(status);

        return new ListTasksResult(tasks, total);
    }
}
